File: file.js
import express from "express";
import cors from "cors";
import { PrismaClient } from "@prisma/client";

const app = express();
const prisma = new PrismaClient();
app.use(cors());
app.use(express.json());

app.get("/languages", async (req, res) => {
  const langs = await prisma.language.findMany();
  res.json(langs);
});

app.post("/progress", async (req, res) => {
  const { userId, lessonId, score } = req.body;
  const progress = await prisma.progress.upsert({
    where: { userId_lessonId: { userId, lessonId } },
    update: { score },
    create: { userId, lessonId, score }
  });
  res.json(progress);
});

app.listen(4000, () => console.log("API running on port 4000"));

File: file.py
def award_xp(user_id, score):
    xp = 0
    if score > 80:
        xp = 20
    elif score > 50:
        xp = 10
    else:
        xp = 5
    update_user_xp(user_id, xp)
    check_badges(user_id)
